### ISSUE TITLE
Replaying replays is incorrect for some replays set before ~2019

### ISSUE THREAD

**Issue by sometimeszzzzzzz** [ppy/osu-stable-issues#554]
Sometime in 2019, the way the game handled replay frames was changed, but there was no exception added for replays set before this change. So playing back those replays on newer clients is off for some replays, specifically the ones where notelock occurs.

Most notable example:
https://osu.ppy.sh/scores/osu/2664504731
[vid of score](https://youtu.be/79yz9TTnMLU)
The score was 95.43% when it was set, but watching the replay now shows 92.91% at the end.

The difference is pretty small for most other replays where this happens e.g.
https://osu.ppy.sh/scores/osu/2255585477 84.98% -> 84.67%
Not much of a difference, but if it's an easy fix, having those replays play back the same way as when they were set would be great.

**Comment by peppy** [comment]
if the reason is the one i'm thinking of, it was a fix for replays recording incorrectly. as such this can't be fixed because the source is wrong.

you're welcome to test the same replays on stable-fallback (download link from the changelog comments section) to confirm.

**Comment by peppy** [comment]
Can you confirm you checked this?

**Comment by sometimeszzzzzzz** [comment]
Sorry, I wasn't too sure about what you meant by the source is wrong.
e.g. for the first score I gave (https://osu.ppy.sh/scores/osu/2664504731), I'm pretty sure that when Ceptin set the play, it showed as 95.43% for him. So I'm not too sure what you mean by the source being wrong.

The play was 95.43% when it was set, and when played back on any client before ~2019 it shows 95.43% (tested on stable-fallback and b20171227.2, [circle people's vid](https://youtu.be/6fBMnnEiMRk) which was probably using 20181212.2 also shows 95.43%).

If this helps, some specific details about what I think is causing this difference:
(I'm not too sure about this though, since this is from my understanding of how replays works, which could be wrong)

Best way I can explain it is with an example:
Two circles, Circle 1 at 1000ms and Circle 2 at 1111ms. The hitwindow50 is 100ms. So Circle 1 will auto miss after 1100ms.
There is a keypress at 1105ms with cursor on Circle 2.

Before the change in 2019, the replay frames would've been something like this:
```
Time | Keypress | x, y coords (those don't matter for this example)
...
1060 | None
1076 | None 
1093 | None
1105 | M1
1110 | M1
...
```
And this would've shown as a hit when played back on a client before the change.

Here's what the replay frames might look like now if the exact same play was done:
```
Time | Keypress | x, y coords
...
1060 | None
1076 | None 
1093 | None
1101 | None
1105 | M1
1110 | M1
...
```
There's an extra frame added so that the game updates and removes Circle 1 after it is missed. And frame handling was changed in a way so that this frame is necessary for Circle 2 to be hit now.

Playing back the replay frames from the first table on newer clients will cause Circle 2 to notelock/shake. At frame 1105, the last time the hit objects were updated was at frame 1093, and Circle 1 still existed then. Because keypresses are handled first, then hit objects are updated, at frame 1105, Circle 1 still exists, so Circle 2 will notelock/shake. Then after that Circle 1 is removed.

This wasn't the case before though. Purely guessing, but maybe there was some kind of exception before to allow Circle 2 to be hit, or maybe hit objects were updated first, then keypresses were handled.

**Comment by peppy** [comment]
Thanks for the additional information. Your thinking is correct, and at the time we decided it wasn't so bad to break replays. I'll take a quick look if fixing this is feasible, and give a response either way.

**Comment by Magnus-Cosmos** [comment]
I think issue #514 is also related to this, where re-watching auto's replay breaks due to how the frames are handled.

**Comment by Flux3on** [comment]
Yeah this happened with a play I did back in 2017. I got 11.89% accuracy but when I play it back the replay shows up as 11.55%

(This was on the big black with nofail when I sucked at the game .3.)

**Comment by peppy** [comment]
Just going through my self-assigned issues, and this doesn't seem like something we will be fixing in stable. If anyone experiencing this able to test if it's still broken in lazer? If so, we can look at getting it fixed over there.
